cpssEnabler/linuxNoKernelModule/drivers/dmaDriver.c

```diff
@@ -28,320 +28,873 @@ disclaimer.
+ loff_t off;
+ int domain;
+ unsigned int bus;
+ unsigned int devfn;
+ int reset_done = 0;
+
+ if (!m->dev) {
+	pr_info("%s: No PCI device assigned!\n", __func__);
+	return -ENOENT;
+ 	}
+
+ for (i=0; i<2; i++) {
+ 	 off = m->pci_offset;
+	 domain = (off >> 16) & 0xffff;
+	 bus = (off >> 8) & 0xff;
+	 devfn = PCI_DEVFN(((off >> 3) & 0x1f), (off & 0x07));
+
+	 pdev = pci_get_domain_bus_and_slot(domain, bus + i, devfn);
+
+	 if (!pdev) {
+		pr_info("%s: Failed to get PCI device %x:%x:%x.%x\n", __func__, domain, bus + i, (unsigned)((off >> 3) & 0x1f), (unsigned)(off & 0x07));
+		continue;
+	 	}
+
+	 m->pdevs_list[i] = pdev;
+	 err = pcim_enable_device(pdev);
+
+	 if (err) {
+		pr_info("%s: pcim enable device failed err %d!\n",__func__,err);
+		return err;
+	 }
+
+        if (!err){
+
+                m->base[i][2] = pcim_iomap(pdev, 2, 64*1024*1024); /* switching registers/MG 64M AC3X/BC2/Aldrin only*/
+
+               if (m->base[i][2]) {
+                       /* 
+                        * Write Receive SDMA Queue Command Register in MG with stop all RX queues Values
+                        * This will disable all of the SDMA RX reception.
+                        * This covers situation in
+                        * which SIGKILL is sent to CPSS application which will cause it to release
+                        * the mvDMA driver file descriptor and call this callback. Writing these
+                        * bits with ones will ensure the Packet Processor SDMA RX is reset to a determined
+                        * state without having the RX DMA running unsupervised.
+                        */
+
+						/* disable retries on resource error */
+						writel(RX_SDMA_ERR_MOD, m->base[i][2] + MG_RX_SDMA_RES_ERR_MOD01);
+			   			
+						writel(RX_SDMA_ERR_MOD, m->base[i][2] + MG_RX_SDMA_RES_ERR_MOD01 + 0x4);
+						
+						for (ii=0; ii<6; ii++)
+							writel(RX_SDMA_ERR_MOD, m->base[i][2] + MG_RX_SDMA_RES_ERR_MOD27 + (0x4 * ii));
+						 
+						 /* Stop SDMA RX - prevent writing on memory which will be released to kernel */
+						writel(STOP_ALL_Q_BITS, m->base[i][2] + MG_RX_SDMA_Q_CMD_REG);
+						
+			   			/* Stop SDMA TX - descriptor writeback might corrupt memory */
+			   		   writel(STOP_ALL_Q_BITS, m->base[i][2] + MG_TX_SDMA_Q_CMD_REG);
+						
+						/* Stop FDB AU messages to CPU (FDB unit) */
+
+						/* Stop AU messages on MG (DMA) */
+					   writel(STOP_AU_Q_BIT, m->base[i][2] + MG_AU_Q_HOST_CONF_REG);
+						
+						/* Reset AU+FU queue counters (reset queue) on MG (DMA) */
+					   writel(RESET_AU_Q_CNTRS_BITS | STOP_AU_Q_BIT, m->base[i][2] + MG_AU_Q_HOST_CONF_REG);
+						
+						/* Finally, completely reset SDMA: */
+						val = readl(m->base[i][2] + MG_EXT_GLOBAL_CNTRL_REG);
+						
+						val |= SDMA_SW_RST_BIT;
+						
+						writel(val, m->base[i][2] + MG_EXT_GLOBAL_CNTRL_REG);
+						
+					    val &= ~SDMA_SW_RST_BIT;
+			   
+					    writel(val, m->base[i][2] + MG_EXT_GLOBAL_CNTRL_REG);
```

Comment:
memory barrier before this to ensure ordering of taking block into and out of reset
memory barrier after this to ensure device is out of reset quickly and before other accesses occur
is there a minimum duration the block needs to be kept in reset, or a bit that needs to be polled to confirm that the state has been entered?

Reply:
No minimum duration nor polling is needed